internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -634,21 +638,26 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 
 		if !cd.Ready {
 			log.Debug("Composed resource is not yet ready", "id", id)
+			unready = append(unready, cd)
 			r.record.Event(xr, event.Normal(reasonCompose, fmt.Sprintf("Composed resource %q is not yet ready", id)))
 			continue
 		}
-
-		ready++
 	}
 
 	xr.SetConditions(xpv1.ReconcileSuccess())
 
 	// TODO(muvaf): If a resource becomes Unavailable at some point, should we
 	// still report it as Creating?
-	if ready != len(res.Composed) {
+	if len(unready) > 0 {
 		// We want to requeue to wait for our composed resources to
 		// become ready, since we can't watch them.
-		xr.SetConditions(xpv1.Creating())
+		names := make([]string, len(unready))
+		for i, cd := range unready {
+			names[i] = cd.ResourceName
```

Comment:
is `ResourceName` guaranteed to be a non empty string? https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/reconciler.go#L628-L633 is making think that it doesn't have that guarantee - would we want to do something different with the name here, e.g. fall back to its index number? 🤔

Reply:
we should print the group kind and name. I was there before actually, then thought that the names are what ppl refer to. Now you are (rightfully) saying they are optional 🙂

Reply:
actually, kind and name are not guaranteed either. Especially, with the Kubernetes provider these are all objects. The index exists in p&f, not in functions afaik. Looks like we need some heuristic decided by the compositor. What matters here is that we find something pragmatic rather than perfect. Would rather do something imperfect than postponing this topic another 2 years. So +1 putting there the best we have to identify the object. This might also encourage ppl to put proper resource names there to avoid "resource 7" like messages.

Reply:
Double checked: I think we are good. For p&f, anonymous resources get `%d` as name. I have made that a little prettier via `resource %d` now. In functions, anonymous resources are not allowed:
> The PTFComposer does not support anonymous templates

@jbw976 ptal

Reply:
sounds good @sttts, thanks for looking deeper into that! ✅